File: clients/android/app/src/main/kotlin/com/lextures/android/features/parent/ConferenceReminderScheduler.kt
package com.lextures.android.features.parent

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import androidx.core.content.getSystemService
import com.lextures.android.R
import com.lextures.android.core.lms.ConferenceSlot
import com.lextures.android.core.lms.DueReminderLeadTime
import com.lextures.android.core.lms.LmsDates
import com.lextures.android.features.planner.DueReminderScheduler
import com.lextures.android.features.planner.PlannerReminderReceiver
import java.time.Instant

/** Schedules local reminders for booked parent–teacher conferences (M10.2 / M0.1). */
object ConferenceReminderScheduler {
    fun scheduleReminder(
        context: Context,
        slot: ConferenceSlot,
        teacherName: String,
        childName: String,
    ) {
        val lead = DueReminderScheduler.selectedLeadTime(context)
        cancelReminder(context, slot.id)
        if (lead == DueReminderLeadTime.None) return
        val start = LmsDates.parse(slot.startAt) ?: return
        val fireAt = start.minusSeconds(lead.minutes.toLong() * 60)
        if (fireAt.isBefore(Instant.now())) return

        val title = context.getString(R.string.mobile_parent_conferences_reminder_title)
        val body = context.getString(
            R.string.mobile_parent_conferences_reminder_body,
            teacherName,
            childName,
        )
        val intent = Intent(context, PlannerReminderReceiver::class.java).apply {
            putExtra(PlannerReminderReceiver.EXTRA_TITLE, title)
            putExtra(PlannerReminderReceiver.EXTRA_BODY, body)
        }
        val pending = PendingIntent.getBroadcast(
            context,
            reminderRequestCode(slot.id),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        context.getSystemService<AlarmManager>()?.setAndAllowWhileIdle(
            AlarmManager.RTC_WAKEUP,
            fireAt.toEpochMilli(),
            pending,
        )
    }

    fun cancelReminder(context: Context, slotId: String) {
        val intent = Intent(context, PlannerReminderReceiver::class.java)
        val pending = PendingIntent.getBroadcast(
            context,
            reminderRequestCode(slotId),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        context.getSystemService<AlarmManager>()?.cancel(pending)
    }

    private fun reminderRequestCode(slotId: String): Int = "conference:$slotId".hashCode()
}

File: clients/ios/Lextures/Features/Planner/ConferenceReminderScheduler.swift
import Foundation
import UserNotifications

/// Schedules local reminders for booked parent–teacher conferences (M10.2 / M0.1).
enum ConferenceReminderScheduler {
    static func notificationIdentifier(for slotId: String) -> String {
        "conference-reminder:\(slotId)"
    }

    static func scheduleReminder(
        slot: ConferenceSlot,
        teacherName: String,
        childName: String
    ) async {
        let lead = DueReminderScheduler.selectedLeadTime
        await cancelReminder(for: slot.id)
        guard lead != .none, let start = LMSDates.parse(slot.startAt) else { return }
        let fireDate = start.addingTimeInterval(-Double(lead.rawValue * 60))
        guard fireDate > Date() else { return }

        let content = UNMutableNotificationContent()
        content.title = L.text("mobile.parent.conferences.reminder.title")
        content.body = L.format(
            "mobile.parent.conferences.reminder.body",
            teacherName,
            childName
        )
        content.sound = .default
        content.userInfo = ["action_url": "/parent/conferences"]

        let components = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute],
            from: fireDate
        )
        let trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
        let request = UNNotificationRequest(
            identifier: notificationIdentifier(for: slot.id),
            content: content,
            trigger: trigger
        )
        try? await UNUserNotificationCenter.current().add(request)
    }

    static func cancelReminder(for slotId: String) async {
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: [notificationIdentifier(for: slotId)])
    }
}
